052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,156 @@
+
+![Current Situation](images/052-current.png)
+
+- `org.apache.kafka.common.metrics.JmxReporter` is the reporter implementation for the Kafka Metrics.
+- `org.apache.kafka.server.metrics.FilteringJmxReporter` is the reporter implementation for the Yammer Metrics. It’s named `FilteringJmxReporter` because it extends the JmxReporter class from the Yammer library and adds an option to select the metrics to report.
+
+
+## Motivation
+
+I propose updating the metrics collection pipeline for the following reasons:
+
+1. The current metrics collection pipeline is pretty convoluted. We have metrics reporters first exposing metrics via JMX before using a Java agent to expose them again via HTTP. Using metrics reporters to directly expose metrics to Prometheus would significantly simplify it by removing JMX from the picture and removing jmx_exporter.
+1. Each component along the pipeline has its own configurations and specificities. For example, since Kafka 3.4.0, it’s possible to disable the JmxReporter for KafkaMetrics by setting [auto.include.jmx.reporter](https://kafka.apache.org/documentation/#brokerconfigs_auto.include.jmx.reporter) to `false`. It is currently not possible to disable `FilteringJmxReporter` (it’s something I plan to address in the near future). Also both jmx_exporter and the built-in reporter allow selecting metrics to collect. 
+1. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. When metrics are renamed, the grafana dashboards we provide can’t be reused. In addition, due to the rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions. 
+1. Recently jmx_exporter was impacted by CVEs in its dependencies (CVE-2022-38752, CVE-2022-25857).
+
+
+## Proposal
+
+The proposal is to build a metrics reporter, `PrometheusMetricsReporter`, that implements both `org.apache.kafka.common.metrics.MetricsReporter` and `kafka.metrics.KafkaMetricsReporter`, and that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
+
+![Proposal](images/052-proposal.png)
+
+To enable the reporter, the following Kafka configurations must be set:
+
+```
+metric.reporters=com.mickaelmaison.PrometheusMetricsReporter # For KafkaMetrics
+kafka.metrics.reporters=com.mickaelmaison.PrometheusMetricsReporter # For YammerMetrics - This is only required for brokers/controllers.
+```
+
+The reporter will expose the following configurations:
+- `prometheus.metrics.reporter.port`: The HTTP port to expose the metrics. Default: `8080`
+- `prometheus.metrics.reporter.allowlist`: A comma separated list of regex patterns to specify the metrics to collect. Default: `.*`. The patterns must match the metric names emitted by the reporter.
```

Comment:
I think the metric names are stable (e.g. tags are in a deterministic order), but it's worth thinking about compatibility here. E.g. what if an existing kafka metric gets a new tag, thus changing the string serialization of its name.